Create event

IvanSibirtsev opened the pull request.

IvanSibirtsev committed:
add photosToTempSaveService

IvanSibirtsev committed:
add Publisher and RabbitMqConfig

IvanSibirtsev committed:
Add Handler and SaveEventResult

IvanSibirtsev committed:
add EventSaveService and RabbitMqMessage

IvanSibirtsev committed:
add LoadFromTempService

IvanSibirtsev committed:
add ImageValidator

IvanSibirtsev committed:
add ImageToDbPreparer

IvanSibirtsev committed:
add PhotosEntity

IvanSibirtsev committed:
add SaveToDbService

IvanSibirtsev committed:
add RabbitMqMessageHandler

IvanSibirtsev committed:
add Consumer

IvanSibirtsev committed:
add Post and GetGuid methods to EventController

IvanSibirtsev committed:
update .csproj

IvanSibirtsev committed:
cleanup

IvanSibirtsev committed:
Merge remote-tracking branch 'origin/dev' into create-event

# Conflicts:
#	Eventnet/Eventnet.Api/Config/ApplicationMappingProfile.cs
#	Eventnet/Eventnet.Api/Controllers/EventController.cs
#	Eventnet/Eventnet.Api/Eventnet.Api.csproj
#	Eventnet/Eventnet.Api/Program.cs

IvanSibirtsev committed:
merge

IvanSibirtsev committed:
add docker-compose for rabbitMq

DaGnYMilleR commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -19,19 +20,25 @@ public class EventController : Controller
+    private readonly IEventSaveService eventSaveService;
     private readonly LinkGenerator linkGenerator;
+    private readonly IMapper mapper;
+    private readonly UserManager<UserEntity> userManager;
```
Field 'userManager' is assigned but its value is never used

DaGnYMilleR replied:
епта

DaGnYMilleR commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -104,10 +113,26 @@ public IActionResult GetEventsByName(string? eventName, [FromQuery(Name = "m")]
     public IActionResult CreateEvent([FromBody] CreateEventModel createModel)
     {
-        throw new NotImplementedException();
+        if (createModel is null)
```
Expression is always false

DaGnYMilleR commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -104,10 +113,26 @@ public IActionResult GetEventsByName(string? eventName, [FromQuery(Name = "m")]
+            return BadRequest();
+        var createdEvent = mapper.Map<Event>(createModel);
+        var files = createModel.Files;
+        eventSaveService.Save(createdEvent, files);
```
Метод же асинхронный. Обычно для них используют постфикс Async
Ну и await, да

DaGnYMilleR commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -104,10 +113,26 @@ public IActionResult GetEventsByName(string? eventName, [FromQuery(Name = "m")]
+        });
+    }
+
+    [HttpGet("isCreated")]
```
kebab

DaGnYMilleR commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -104,10 +113,26 @@ public IActionResult GetEventsByName(string? eventName, [FromQuery(Name = "m")]
+        eventSaveService.Save(createdEvent, files);
+        return Ok(new
+        {
+            StatusCode = StatusCodes.Status202Accepted
```
можно заменить на return Accepted();

DaGnYMilleR commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -129,6 +154,13 @@ public async Task<IActionResult> DeleteEvent(Guid eventId)
     }
+    
+    [HttpGet("guid")]
+    public IActionResult GenerateEventGuid()
```
вкусовщина, но я бы предпочел inline, хотя, зная наших любителей писать скобки вместо лямбд...

gODeaLoAple replied:
В контроллерах только statement-body!

DaGnYMilleR commented on Eventnet/Eventnet.Api/Config/ApplicationMappingProfile.cs:
```diff
@@ -11,6 +11,9 @@ public ApplicationMappingProfile()
         CreateMap<EventEntity, Event>();
         CreateMap<EventEntity, EventNameModel>();
         CreateMap<LocationEntity, Location>();
+        CreateMap<Event, CreateEventModel>()
```
Ниче не понял, что ты тут хотел сделать. Но почему мы ставим юзеру новый id?

DaGnYMilleR commented on Eventnet/Eventnet.Api/Eventnet.Api.csproj:
```diff
@@ -9,25 +9,30 @@
-        <PackageReference Include="X.PagedList.Mvc.Core" Version="8.1.0"/>
+        <PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="6.0.2" />
+        <PackageReference Include="PagedList" Version="1.17.0" />
+        <PackageReference Include="PagedList.Mvc" Version="4.5.0" />
```
![image](https://user-images.githubusercontent.com/37438377/159059885-0eb5eb67-9144-4677-b507-22e688895ec9.png)
чета с либами не то...

DaGnYMilleR replied:
________(или это в другой ветке зафикшено?)

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/IPhotosToTempSaveService.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace Eventnet.Services;
+
+public interface IPhotosToTempSaveService
```
to temp что? кажется не хватает folder

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/IEventSaveService.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace Eventnet.Services;
+
+public interface IEventSaveService
```
Общий коммент. Стоит поместить эти сервисы в отдельную подпапку, шоб у нас каша не получалась

DaGnYMilleR commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -104,10 +113,26 @@ public IActionResult GetEventsByName(string? eventName, [FromQuery(Name = "m")]
     }
 
-    [HttpPost("create")]
+    [HttpPost]
```
надо бы [Authorize] навесить

DaGnYMilleR commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -104,10 +113,26 @@ public IActionResult GetEventsByName(string? eventName, [FromQuery(Name = "m")]
-        throw new NotImplementedException();
+        if (createModel is null)
+            return BadRequest();
+        var createdEvent = mapper.Map<Event>(createModel);
```
А откуда тут ваще id для евента возьмется? Возможно тогда это просто опечатка в маппере и там должно быть не UserId?

DaGnYMilleR commented on Eventnet/Eventnet.Api/Models/CreateEventModel.cs:
```diff
@@ -6,4 +6,5 @@ public record CreateEventModel(
     DateTime? EndAt,
-    Location Location);
\ No newline at end of file
+    Location Location,
```
Не могу оставить коммент выше. Пользователь может создавать событие только для себя самого? Если да то стоит id пользователя получать через CurrentUserService(дописать там метод для его получения) который у меня лежит в еще на данный момент не вмерженном пре

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/EventSaveService.cs:
```diff
@@ -0,0 +1,38 @@
+
+    public async void Save(Event savedEvent, IFormFile[] photos)
+    {
+        var path = photosToTempSaveService.SaveToTemp(savedEvent.Id, photos);
```
кажется, что мы можем хотеть от фотки не просто ее наличие, но и какуюнить доп инфу(например, размер, расширение etc). Давайте обсудим.

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/Handler.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace Eventnet.Services;
+
+public class Handler
```
кажется, что имя handler не совсем полностью отражает суть...

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/Handler.cs:
```diff
@@ -0,0 +1,22 @@
+    public SaveEventResult GetValue(Guid id)
+    {
+        saveEventInformation.Remove(id, out var value);
+        return value;
```
Possible null reference return.

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/Handler.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    private readonly ConcurrentDictionary<Guid, SaveEventResult> saveEventInformation = new();
+
+    public void Update(Guid id, SaveEventResult value)
```
ну этот метод же не только Update но и Add делает...

gODeaLoAple replied:
В EFCore такой же нейминг. Считаю, что можно просто Update оставить.

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/Handler.cs:
```diff
@@ -0,0 +1,22 @@
+
+    public void Update(Guid id, SaveEventResult value)
+    {
+        // может ли возникнуть ситуация когда нужно будет использовать TryUpdate ?
```
кажется, что лучше не *** мозга, а заюзать AddOrUpdate

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/PublishEventService.cs:
```diff
@@ -0,0 +1,29 @@
+    {
+        queue = config.Queue;
+        var factory = new ConnectionFactory { HostName = config.HostName };
+        var connection = factory.CreateConnection();
```
connection и channel IDisposable

DaGnYMilleR commented on Eventnet/Eventnet.Infrastructure/SaveToDbService.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+        this.preparer = preparer;
+        this.mapper = mapper;
+        dbContext = factory.CreateScope().ServiceProvider.GetRequiredService<ApplicationDbContext>();
```
А в чем проблема сразу ApplicationDbContext прокинуть?

DaGnYMilleR commented on Eventnet/Eventnet.Infrastructure/ImageServices/ILoadFromTempService.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace Eventnet.Infrastructure.ImageServices;
+
+public interface ILoadFromTempService
```
А почему SaveToTemp в одном проекте, а Load в другом?

DaGnYMilleR commented on Eventnet/Eventnet.Infrastructure/ImageServices/LoadFromTempService.cs:
```diff
@@ -0,0 +1,19 @@
+        foreach (var filename in filenames)
+        {
+            images.Add(Image.FromFile(filename));
+            File.Delete(filename); // возможно, их стоит удалять в другой момент
```
Кажется. что да. Потому что при загрузке фоток совсем не ожидаешь, что они пропадут)

gODeaLoAple replied:
Согласен, удалять надо только в случае успешного перемещения.

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/ConsumeEventService.cs:
```diff
@@ -0,0 +1,48 @@
+        {
+            var content = Encoding.UTF8.GetString(ea.Body.ToArray());
+            var message = JsonSerializer.Deserialize<RabbitMqMessage>(content);
+            rabbitMqMessageHandler.Handle(message);
```
Possible null reference argument for parameter 'rabbitMqMessage' in 'Eventnet.Services.IRabbitMqMessageHandler.Handle'.

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/Handler.cs:
```diff
@@ -0,0 +1,22 @@
+        return value;
+    }
+
+    public bool ContainsInformationAbout(Guid id) => saveEventInformation.ContainsKey(id);
```
Кажется, что тут может что-то пойти не так, так как ты сначала проверяешь, что ключ есть, а затем берешь значение. Между этими двумя действиями может вклиниться какой-нить другой поток, которому Handler тоже ответит true, и удалить раньше. TryGetValue тут подойдет наверн

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/PhotosToTempSaveService.cs:
```diff
@@ -0,0 +1,25 @@
+        {
+            if (file.Length <= 0)
+                continue;
+            var filePath = Path.Combine(eventPhotosPath, file.FileName);
```
Что будет, если файл с таким именем уже есть в temp? Если требуется сохранить имя исходного файла, то где-то выше я уже писал, о том что нужна отдельная модель для фоток, так вот, это можно туда же

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/PhotosToTempSaveService.cs:
```diff
@@ -0,0 +1,25 @@
+        var pathToPhotos = "temp"; // TODO: вынести в кофниг, когда он повяится
+        if (!Directory.Exists(pathToPhotos))
+            Directory.CreateDirectory(pathToPhotos);
+        var eventPhotosPath = Path.Combine(pathToPhotos, id.ToString());
```
Постой брат. Мы создаем для каждого заявления отдельную папку? С тем учетом, что у нас в среднем для каждого события будет 1 фотка(личные ощущения)? Так как ты возвращаешь путь до файла, то можно его генерить через Guid или Path.GetRandomFileName(). Надо обсудить.

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/RabbitMqMessageHandler.cs:
```diff
@@ -0,0 +1,44 @@
+        try
+        {
+            var photos = loadFromTempService.LoadImages(rabbitMqMessage.PathToPhotos);
+            if (validator.Validate(photos, out exception))
```
А разве нам не нужно валидировать данные самого события? Имя там, даты и т.п.

DaGnYMilleR commented on Eventnet/Eventnet.DataAccess/PhotosEntity.cs:
```diff
@@ -0,0 +1,13 @@
+
+public class PhotosEntity
+{
+    public string pathToPhotos { get; set; }
```
Свойство с маленькой строки? -_-

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/RabbitMqConfig.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace Eventnet.Services;
+
+public class RabbitMqConfig
```
Стоит читать хотяб из appsettings.json наверн, пока мы не прикрутили нормальные переменные среды

DaGnYMilleR commented on Eventnet/Eventnet.DataAccess/ApplicationDbContext.cs:
```diff
@@ -9,7 +9,8 @@ namespace Eventnet.DataAccess;
 {
     public DbSet<EventEntity> Events { get; set; }
-    
+    public DbSet<PhotosEntity> Photos { get; set; }
```
А чо где миграция?

DaGnYMilleR commented on Eventnet/Eventnet.Domain/Eventnet.Domain.csproj:
```diff
@@ -7,12 +7,12 @@
 
     <ItemGroup>
-        <ProjectReference Include="..\Eventnet.DataAccess\Eventnet.DataAccess.csproj"/>
+        <ProjectReference Include="..\Eventnet.DataAccess\Eventnet.DataAccess.csproj" />
```
Domain зависит от DataAccess? Точно нет, проходили в PathFindere такое....

DaGnYMilleR commented on Eventnet/Eventnet.Infrastructure/ImageServices/IImageToDbPreparer.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace Eventnet.Infrastructure.ImageServices;
+
+public interface IImageToDbPreparer
```
Preparer и метод Save. Го обсудим

gODeaLoAple replied:
Плохой нейминг.

DaGnYMilleR reviewed (changes requested):
пока subject не jopa - не приму

Хочу добавить пока не забыл. Кажется, что хранить все в апи и инфре не оч идея, надо сесть, обсудить и определиться с предметной областью(либо выпилить domain к чертям)

gODeaLoAple commented on Eventnet/Eventnet.Api/Services/EventSaveService.cs:
```diff
@@ -0,0 +1,38 @@
+        this.handler = handler;
+    }
+
+    public async void Save(Event savedEvent, IFormFile[] photos)
```
```suggestion
    public async void Save(Event eventForSave, IFormFile[] photos)
```

gODeaLoAple commented on Eventnet/Eventnet.Api/Services/Handler.cs:
```diff
@@ -0,0 +1,22 @@
+﻿using System.Collections.Concurrent;
```
Кажется, что имеет смысл вынести либо в домейн, либо в инфру, т.к. по факту это просто кеш.

gODeaLoAple commented on Eventnet/Eventnet.Api/Services/RabbitMqMessageHandler.cs:
```diff
@@ -0,0 +1,44 @@
+        try
+        {
+            var photos = loadFromTempService.LoadImages(rabbitMqMessage.PathToPhotos);
+            if (validator.Validate(photos, out exception))
```
Почему Validate имеет такой интерфейс? Может, стоит кидать ошибку, если данные неконсистенты? 
Для этого можно просто создать какой-нибудь EventCreationValidationException и либо наследоваться от него, либо просто в него сообщение пробрасывать. out параметр выглядит лишним.

DaGnYMilleR replied:
Как альтернативу бросанию ошибки предложу возвращать какой нить ValidationResult

gODeaLoAple replied:
Че мелочиться, подрубай монады!

gODeaLoAple commented on Eventnet/Eventnet.Api/Services/RabbitMqMessageHandler.cs:
```diff
@@ -0,0 +1,44 @@
+            var photos = loadFromTempService.LoadImages(rabbitMqMessage.PathToPhotos);
+            if (validator.Validate(photos, out exception))
+            {
+                saveToDbService.SaveImages(photos, id);
```
Даже тут надо сначала сохранить ивент, а потом изображения, потому что на у фоток будет foreign key на ивент.

gODeaLoAple commented on Eventnet/Eventnet.Api/Services/RabbitMqMessageHandler.cs:
```diff
@@ -0,0 +1,44 @@
+            if (validator.Validate(photos, out exception))
+            {
+                saveToDbService.SaveImages(photos, id);
+                saveToDbService.SaveEvent(rabbitMqMessage);
```
Почему saveToDbService принимает rabbitMqMessage?

gODeaLoAple commented on Eventnet/Eventnet.DataAccess/PhotosEntity.cs:
```diff
@@ -0,0 +1,13 @@
+﻿namespace Eventnet.DataAccess;
```
Доработать модель. Сделать более консистентной. 
1. ForeignKey -> Event
2. Create configuration

gODeaLoAple commented on Eventnet/Eventnet.Infrastructure/ISaveToDbService.cs:
```diff
@@ -0,0 +1,9 @@
+public interface ISaveToDbService
+{
+    void SaveImages(List<Image> images, Guid id);
+    void SaveEvent(RabbitMqMessage message);
```
SaveEvent, но принимает RabbitMqMessage

gODeaLoAple commented on Eventnet/Eventnet.Infrastructure/ISaveToDbService.cs:
```diff
@@ -0,0 +1,9 @@
+
+public interface ISaveToDbService
+{
+    void SaveImages(List<Image> images, Guid id);
```
Непонятно, к чему относится id. Возможно исчезнет после переосмысления модели Image в бд.

gODeaLoAple commented on Eventnet/Eventnet.Infrastructure/ImageServices/IImageToDbPreparer.cs:
```diff
@@ -0,0 +1,8 @@
+
+public interface IImageToDbPreparer
+{
+    string Save(List<Image> images, Guid id);
```
Странный интерфейс и название.

gODeaLoAple commented on Eventnet/Eventnet.Infrastructure/ImageServices/ImageToDbPreparer.cs:
```diff
@@ -0,0 +1,24 @@
+        {
+            var path = Path.Combine(dir, i + ".jpeg");
+            var image = images[i];
+            image.Save(path, ImageFormat.Jpeg);
```
Нет обработки ошибок. Возможно, в случае ошибок надо возвращать файлы, которыене удалось сохранить.

gODeaLoAple commented on Eventnet/Eventnet.Infrastructure/SaveToDbService.cs:
```diff
@@ -0,0 +1,34 @@
+    public void SaveEvent(RabbitMqMessage message)
+    {
+        var eventEntity = mapper.Map<EventEntity>(message.Event);
+        dbContext.Events.Add(eventEntity);
```
Ты уверен, что сохраняешь event?

gODeaLoAple commented on Eventnet/Eventnet.Infrastructure/SaveToDbService.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+        var path = preparer.Save(images, id);
+        var photosEntity = new PhotosEntity(path, id);
+        dbContext.Photos.Add(photosEntity);
```
Ты уверен, что сохраняешь photo?

gODeaLoAple reviewed (changes requested):
Нет тестов - нет аппрува. Без негатива.

IvanSibirtsev committed:
kebab

IvanSibirtsev committed:
fix to return accepted

IvanSibirtsev committed:
made IEventSaveService method Save async

IvanSibirtsev committed:
add [Authorize]

IvanSibirtsev committed:
change CreateEventModel and Automapper config:
1. add Guid id
2. rename Files -> Photos

IvanSibirtsev committed:
move services to infrastructure

IvanSibirtsev committed:
fix delete photos from temp

IvanSibirtsev committed:
renaming. Temp -> TempFolder

IvanSibirtsev committed:
fix SaveToDbService.SaveEvent signature. RabbitMqMessage -> Event

IvanSibirtsev committed:
fix handler

IvanSibirtsev committed:
make RabbitMQ to send image not path to image

IvanSibirtsev committed:
delete unused classes

IvanSibirtsev committed:
add PhotoEntity, PhotoConfiguration and migrations

IvanSibirtsev committed:
rename ImageToDbPreparer -> PhotoToStorageSaveService
fix SaveToDbService

IvanSibirtsev committed:
move IPublishEventService and IEventSaveService to SaveServices folder

IvanSibirtsev committed:
add Dispose to PublishEventService

IvanSibirtsev committed:
rename ImageServices -> PhotoServices, ImageValidator -> PhotoValidator

IvanSibirtsev committed:
add EventValidator, PhotoValidator and EventCreationValidator

DaGnYMilleR commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -19,19 +21,25 @@ public class EventController : Controller
     private readonly IEventFilterMapper filterMapper;
     private readonly ApplicationDbContext dbContext;
-    private readonly IMapper mapper;
+    private readonly IEventSaveService eventSaveService;
```
![image](https://user-images.githubusercontent.com/37438377/161816134-5fd8b8d8-dd1e-4bb1-80f2-6266b830a641.png)

DaGnYMilleR commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -129,6 +151,14 @@ public async Task<IActionResult> DeleteEvent(Guid eventId)
         return Ok(new { eventId });
     }
+    
+    [HttpGet("guid")]
```
Зачем этот эндпоинт нужен?(я задавал этот вопрос, ответа не было, теперь придется)
Почему бы из create-event id не возвращать?

DaGnYMilleR commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -104,10 +114,22 @@ public IActionResult GetEventsByName(string? eventName, [FromQuery(Name = "m")]
+    public IActionResult IsCreated(Guid id)
+    {
+        if (!eventSaveService.IsEventSaved(id, out var exception))
+            return BadRequest(exception);
```
А почему кидаем BadRequest? У нас же может быть такое, что событие еще не создано, а это вроде как не исключительная ситуация

DaGnYMilleR commented on Eventnet/Eventnet.Api/Models/CreateEventModel.cs:
```diff
@@ -1,9 +1,11 @@
+    Guid Id,
+    string OwnerId,
+    DateTime StartDate,
+    DateTime? EndDate,
```
А почему у события может не быть конца?

IvanSibirtsev replied:
тут такое устроили. Итог: я не знаю, давай оставим это и не будем к этому возвращаться

IvanSibirtsev replied:
Ребята не стоит вскрывать эту тему. Вы молодые, шутливые, вам все легко. Это не то. Это не Чикатило и даже не архивы спецслужб. Сюда лучше не лезть. Серьезно, любой из вас будет жалеть. Лучше закройте тему и забудьте что тут писалось. Я вполне понимаю что данным сообщением вызову дополнительный интерес, но хочу сразу предостеречь пытливых - стоп. Остальных просто не найдут.

DaGnYMilleR commented on Eventnet/Eventnet.Infrastructure/Handler.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace Eventnet.Infrastructure;
+
+public class Handler
```
Чому handler то?

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/SaveServices/RabbitMqConfig.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace Eventnet.Services.SaveServices;
+
+public class RabbitMqConfig
```
Ну это же конфиг, почему он в сервисах?
Да и лучше читать это дело из appsettings.json

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/SaveServices/EventSaveService.cs:
```diff
@@ -0,0 +1,47 @@
+        return saveEventResult.IsSaved;
+    }
+
+    private async Task<List<byte[]>> GetStreamsAsync(IFormFile[] photos)
```
static?

DaGnYMilleR commented on Eventnet/Eventnet.DataAccess/ApplicationDbContext.cs:
```diff
@@ -9,7 +9,8 @@ namespace Eventnet.DataAccess;
 {
     public DbSet<EventEntity> Events { get; set; }
-    
+    public DbSet<PhotoEntity> Photos { get; set; }
```
Так тыж не зааплаил конфигурацию...
builder.ApplyConfiguration(new EventConfiguration());

DaGnYMilleR commented on Eventnet/Eventnet.Domain/Eventnet.Domain.csproj:
```diff
@@ -7,12 +7,12 @@
     </PropertyGroup>
```
коммент к файлам Event и Location, там неймспейс говна

DaGnYMilleR commented on Eventnet/Eventnet.Infrastructure/PhotoServices/PhotoToStorageSaveService.cs:
```diff
@@ -0,0 +1,21 @@
+
+public class PhotoToStorageSaveService : IPhotoToStorageSaveService
+{
+    private const string DirToSave = "storage";
```
Путь до файла с фотками лучше из конфига читать

DaGnYMilleR commented on Eventnet/Eventnet.Infrastructure/Validators/EventCreationValidator.cs:
```diff
@@ -0,0 +1,27 @@
+            photoValidationResult && eventValidationResult);
+    }
+
+    private string FormatException(string photoException, string eventException) => (photoException + " " + eventException).Trim();
```
static

DaGnYMilleR commented on Eventnet/Eventnet.Infrastructure/Validators/EventCreationValidationResult.cs:
```diff
@@ -0,0 +1,3 @@
+﻿namespace Eventnet.Infrastructure.Validators;
+
+public record EventCreationValidationResult(string Exception, bool IsOk);
```
IsOk лучше сделать первым параметром и в Exception(лучше назвать типа ErrorMessage btw) сделать по умолчанию = string.Empty

DaGnYMilleR commented on Eventnet/Eventnet.Infrastructure/Validators/PhotoValidator.cs:
```diff
@@ -0,0 +1,10 @@
+public class PhotoValidator : IPhotoValidator
+{
+    public EventCreationValidationResult Validate(List<Image> photos) =>
+        // надо придумать что проверять на бэкенде.
```
Ну, размер хотим проверять, что б нам earth.bmp не закинули

IvanSibirtsev committed:
fix using and delete userManager

IvanSibirtsev committed:
add EventSaveStatus, refactor IsCreated endpoint

IvanSibirtsev committed:
read RabbitMqConfig values from appsettings.json

IvanSibirtsev committed:
add size of images validation

IvanSibirtsev committed:
flip IsOk and ErrorMessage. "" -> string.Empty

IvanSibirtsev committed:
add static to signature

IvanSibirtsev committed:
read PhotoStorageConfig values from appsettings.json

IvanSibirtsev committed:
delete SaveEventResult from Handler

IvanSibirtsev committed:
1. add IConsumeEventService and realisation
2. Rename ConsumeEventService -> BackgroundConsumeEventService

IvanSibirtsev committed:
make IPublishEventService IDisposable

IvanSibirtsev committed:
fix Integration tests. Add Mocks to Consume and Publish Event Services

gODeaLoAple commented on Eventnet/Eventnet.Api/Services/SaveServices/ConsumeEventService.cs:
```diff
@@ -4,45 +4,39 @@
     {
-        stoppingToken.ThrowIfCancellationRequested();
         var consumer = new EventingBasicConsumer(channel);
         consumer.Received += (ch, ea) =>
```
```suggestion
        consumer.Received += (_, ea) =>
```
  <Eventnet.Api>\Services\SaveServices\ConsumeEventService.cs:813 Parameter 'ch' is never used

gODeaLoAple commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -114,25 +117,32 @@ public IActionResult GetEventsByName(string? eventName, [FromQuery(Name = "m")]
-        });
+            return BadRequest("Too large images. Recommended size of all images is 128Mb.");
+        }
+        await eventSaveService.SaveAsync(createdEvent, photos);
```
Заренеймить нормально.

gODeaLoAple commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -114,25 +117,32 @@ public IActionResult GetEventsByName(string? eventName, [FromQuery(Name = "m")]
+        };
     }
 
     // TODO use format https://datatracker.ietf.org/doc/html/rfc6902
```
А почему нет редактирования? Ну, то есть update-a события

gODeaLoAple commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -154,8 +164,9 @@ public async Task<IActionResult> DeleteEvent(Guid eventId)
     }
-    
+
     [HttpGet("guid")]
```
```suggestion
    [HttpGet("request-event-creation")]
```
Может быть так?

DaGnYMilleR replied:
Только давайте не будем здесь id возвращать, пожалуйста, мне прост кажется, что получать id где-то отдельно от создания события не очень хорошо....

gODeaLoAple commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -114,25 +117,32 @@ public IActionResult GetEventsByName(string? eventName, [FromQuery(Name = "m")]
         {
-            StatusCode = StatusCodes.Status202Accepted
-        });
+            return BadRequest("Too large images. Recommended size of all images is 128Mb.");
```
Почему размер захардкожен?

gODeaLoAple commented on Eventnet/Eventnet.Api/Services/SaveServices/IPublishEventService.cs:
```diff
@@ -0,0 +1,6 @@
+
+public interface IPublishEventService : IDisposable
+{
+    Task SendAsync(string message);
```
Ну раз это IPublishService, то пусть функция и называется соответствующе?
```suggestion
    Task PublishAsync(string message);
```

gODeaLoAple commented on Eventnet/Eventnet.DataAccess/PhotoEntity.cs:
```diff
@@ -0,0 +1,14 @@
+
+public class PhotoEntity
+{
+    public Guid Id { get; set; }
```
```suggestion
    public Guid Id { get; private set; }
```

gODeaLoAple commented on Eventnet/Eventnet.DataAccess/PhotoEntity.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public Guid Id { get; set; }
+    
+    public Guid EventId { get; set; }
```
```suggestion
    public Guid EventId { get; }
```

gODeaLoAple commented on Eventnet/Eventnet.Infrastructure/PhotoServices/PhotoToStorageSaveService.cs:
```diff
@@ -0,0 +1,22 @@
+
+    public void Save(Image photo, Guid photoId)
+    {
+        var path = Path.Combine(dirToSave, photoId + ".jpeg");
```
Почему jpeg? Почему захардкожено?

gODeaLoAple commented on Eventnet/Eventnet.Infrastructure/RabbitMqMessageHandler.cs:
```diff
@@ -0,0 +1,53 @@
+        {
+            errorMessage = "Something went wrong on server. Please try again later" + "\n" + errorMessage;
+            status = EventSaveStatus.NotSavedDueToServerError;
+            Console.WriteLine(e);
```
Надо больше console.log

gODeaLoAple commented on Eventnet/Eventnet.Infrastructure/SaveToDbService.cs:
```diff
@@ -1,34 +1,48 @@
+    private async Task SavePhotoToDbAsync(PhotoEntity photoEntity)
     {
-        var eventEntity = mapper.Map<EventEntity>(message.Event);
+        await using var dbContext = factory.CreateScope().ServiceProvider.GetRequiredService<ApplicationDbContext>();
```
А почему ты не хочешь контекст прокидывать в конструктор? Можно же сервис тоже сделать скопед. Либо раз уж на то пошло, лучше инджектить ContextFactory, а не IServiceScopeFactory.

gODeaLoAple commented on Eventnet/Eventnet.Infrastructure/Validators/EventCreationValidationResult.cs:
```diff
@@ -0,0 +1,3 @@
+﻿namespace Eventnet.Infrastructure.Validators;
+
+public record EventCreationValidationResult(bool IsOk, string ErrorMessage);
```
Мне кажется, пора уже добавить Either куда-нибудь в инфраструктуру и использовать его.

gODeaLoAple commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -114,25 +117,32 @@ public IActionResult GetEventsByName(string? eventName, [FromQuery(Name = "m")]
+            return BadRequest("Too large images. Recommended size of all images is 128Mb.");
+        }
+        await eventSaveService.SaveAsync(createdEvent, photos);
+        return Accepted();
```
А ведь если пользователь не прикрепил ни одной фотографии, то событие может быть сохранено быстро, и ответ можно отправить сразу же. Просто вкинул идею на подумать.

gODeaLoAple reviewed (changes requested):
На самом деле большинство сервисов могут быть зарегистрированы Scoped, тогда не будет проблем с инджектом DbContext. Синглтон нужен в основном для хандлеров.

Ну и да. Нет тестов - нет аппрува.

gODeaLoAple commented on Eventnet/Eventnet.Infrastructure/RabbitMqMessageHandler.cs:
```diff
@@ -0,0 +1,53 @@
+            var photos = GetPhotos(binaryPhotos);
+            var result = validator.Validate(photos, eventForSave);
+            errorMessage = result.ErrorMessage;
+            if (result.IsOk)
+            {
+                saveToDbService.SaveEvent(eventForSave);
+                saveToDbService.SavePhotos(photos, id);
+            }
+
+            status = result.IsOk ? EventSaveStatus.Saved : EventSaveStatus.NotSavedDueToUserError;
```
Вкусовщина, но...
```suggestion
            if (result.IsOk)
            {
                saveToDbService.SaveEvent(eventForSave);
                saveToDbService.SavePhotos(photos, id);
                status = EventSaveStatus.Saved;
            }
            else
            {
                status = EventSaveStatus.NotSavedDueToUserError;
            }
```

IvanSibirtsev committed:
Rename SendAsync -> PublishAsync

IvanSibirtsev committed:
1. Create domain class Photo
2. Image -> Photo
3. Photo saved with right format

IvanSibirtsev committed:
1. Rename RecommendedMessageSize -> RecommendedMessageSizeInBytes
2. Add recommendedSizeInMb calculation for BadRequest message

IvanSibirtsev committed:
use IMemoryCache in Handler

IvanSibirtsev committed:
add check for content type of files

IvanSibirtsev committed:
fix FromBody -> FromForm

IvanSibirtsev committed:
add AbstractTestBase and TestWithRabbitMqBase

IvanSibirtsev committed:
add RabbitMqTestsFactory

IvanSibirtsev committed:
add integration tests

IvanSibirtsev committed:
refactor integration tests

IvanSibirtsev committed:
fix PhotoEntity + add new migration

IvanSibirtsev committed:
rename endpoint guid -> request-event-creation"

gODeaLoAple commented on Eventnet/Eventnet.Api.IntegrationTests/EventControllerTests/CreateEventTests/CreateEventShould.cs:
```diff
@@ -0,0 +1,108 @@
+        var multipart = GetEventCreationRequestMessage(eventId, ownerId, fileStream, mediaType);
+        var uri = new UriBuilder(Configuration.BaseUrl) { Path = BaseRoute }.Uri;
+        return await client.PostAsync(uri, multipart);
+    }
```
Некритично, но я бы все вспомогательные методы перемещал вниз.

gODeaLoAple commented on Eventnet/Eventnet.Api.IntegrationTests/EventControllerTests/CreateEventTests/CreateEventShould.cs:
```diff
@@ -0,0 +1,108 @@
+
+namespace Eventnet.Api.IntegrationTests.EventControllerTests.CreateEventTests;
+
+public class CreateEventTestsShould : CreateEventTestsBase
```
Не хватило теста на повторное сохранение того же события. По идее, должно отклонить, либо полностью заменить, но не сохранить оба.

gODeaLoAple replied:
И хочется убедиться, что фотография сохранится.

gODeaLoAple replied:
Желательно проверить, что путь из бд будет указывать на реально существующий файл.

gODeaLoAple commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -118,14 +118,22 @@ public IActionResult GetEventsByName(string? eventName, [FromQuery(Name = "m")]
         }
+        
+        var createdEvent = mapper.Map<Event>(createModel);
         await eventSaveService.SaveAsync(createdEvent, photos);
```
Забыл переименовать

gODeaLoAple commented on Eventnet/Eventnet.DataAccess/Configurations/PhotoConfiguration.cs:
```diff
@@ -13,6 +13,6 @@ public void Configure(EntityTypeBuilder<PhotoEntity> builder)
             .IsRequired()
             .OnDelete(DeleteBehavior.Cascade);
-        builder.Property(x => x.EventId);
+            // builder.Property(x => x.EventId);
```
А вот это надо бы удалить, а не закомментить

gODeaLoAple replied:
Удалить, а не закомментить.

gODeaLoAple commented on Eventnet/Eventnet.Domain/Photo.cs:
```diff
@@ -0,0 +1,35 @@
+            "image/jpeg" => (ImageFormat.Jpeg, ".jpeg"),
+            "image/png"  => (ImageFormat.Png, ".png"),
+            "image/bmp"  => (ImageFormat.Bmp, ".bmp"),
+            _            => throw new ArgumentOutOfRangeException($"Unknown content type {contentType}")
```
Возможно, имело бы смысл добавить кастомную ошибку формата изображения. UnsupportedImageFileException, например

gODeaLoAple replied:
Возможно, стоит использовать кастомный exception. UnsupportedImageFormat.

gODeaLoAple commented on Eventnet/Eventnet.Api/Services/SaveServices/EventSaveService.cs:
```diff
@@ -17,11 +17,19 @@ public EventSaveService(IPublishEventService publishEventService, Handler handle
+        }
+        catch (Exception e)
+        {
+            Console.WriteLine(e);
```
Такое неодобряем. Надо Васю запрячь добавить логирование

gODeaLoAple replied:
x2

gODeaLoAple commented on Eventnet/Eventnet.DataAccess/Migrations/ApplicationDbContextModelSnapshot.cs:
```diff
@@ -33,7 +33,7 @@ protected override void BuildModel(ModelBuilder modelBuilder)
 
                     b.Property<DateTime?>("EndDate")
-                        .HasColumnType("timestamp with time zone");
+                        .HasColumnType("timestamp without time zone");
```
Интересно, как это получилось. наверное, нам нужны таймзоны

gODeaLoAple commented on Eventnet/Eventnet.Infrastructure/Validators/EventCreationValidator.cs:
```diff
@@ -14,14 +14,15 @@ public EventCreationValidator(IPhotoValidator photoValidator, IEventValidator ev
 
-    private static string FormatException(string photoErrorMessage, string eventErrorMessage) => (photoErrorMessage + " " + eventErrorMessage).Trim();
+    private static string FormatException(string photoErrorMessage, string eventErrorMessage) =>
+        (photoErrorMessage + " " + eventErrorMessage).Trim();
```
Ну это кал, конечно, но пока плевать.

gODeaLoAple reviewed (changes requested):
Доработать некоторые моменты, в остальном все нормально

IvanSibirtsev committed:
fix handler.
1. Add ContainsGuid method
2. TryGetValue now don't delete values
3. Now cache saves value for 1 minute

IvanSibirtsev committed:
fix EventSaveService
1. update handler on server error
2. add ContainsGuid method

IvanSibirtsev committed:
add check for two-times provided same EventId

IvanSibirtsev committed:
1. refactoring
2. add new test for two-times provided same EventId

IvanSibirtsev committed:
delete commented line

gODeaLoAple commented on Eventnet/Eventnet.Api.IntegrationTests/EventControllerTests/CreateEventTests/CreateEventShould.cs:
```diff
@@ -46,13 +47,17 @@ public async Task ResponseCode400_WhenFileTypeIsNotSupported()
         
         response.StatusCode.Should().Be(HttpStatusCode.BadRequest);
```
Коммент к названию класса. В интеграционных тестах нейминг такой: {MethodName}Should
И каждый метод - Response{Code}[_When{Condition}[_And{Condition}]...]. 
короче, просто убери "Tests" из названия :)

gODeaLoAple commented on Eventnet/Eventnet.Api.IntegrationTests/EventControllerTests/CreateEventTests/CreateEventShould.cs:
```diff
@@ -105,4 +112,12 @@ public async Task IsCreatedResponseCode400_WhenUnknownGuid()
         response.StatusCode.Should().Be(HttpStatusCode.BadRequest);
     }
+    
+    private async Task<HttpResponseMessage> PostAsync(Guid eventId, Guid ownerId, FileStream fileStream, string mediaType)
```
[Optional] Было бы неплохо заренеймить во что-то, что более четко отражало суть метода. Типа RequestSave или в таком духе.

gODeaLoAple commented on Eventnet/Eventnet.Api/Services/SaveServices/EventSaveService.cs:
```diff
@@ -17,21 +17,25 @@ public EventSaveService(IPublishEventService publishEventService, Handler handle
         }
     }
 
+    public bool ContainsGuid(Guid id) => handler.ContainsGuid(id);
+
```
Плохой нейминг. В одних методах уточняешь, что сущность относится к Event (GetSaveEventResult), в других - считаешь так по-умолчанию (SaveAsync, ContainsGuid).

```suggestion
    public bool IsSaving(Guid eventId) => handler.ContainsGuid(id);

```

gODeaLoAple commented on Eventnet/Eventnet.Infrastructure/Handler.cs:
```diff
@@ -10,17 +10,13 @@ public Handler(IMemoryCache cache)
         this.cache = cache;
     }
+    
+    public bool ContainsGuid(Guid id) => cache.TryGetValue(id, out _);
```
См. коммент к EventSaveService

gODeaLoAple reviewed (approved):
В принципе, можно мержить, но лучше бы немного причесать.

IvanSibirtsev committed:
renamings

IvanSibirtsev committed:
add TODOs to logger task

gODeaLoAple commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -134,13 +134,13 @@ public async Task<IActionResult> CreateEvent([FromForm] CreateEventModel createM
 
         var isSaved = await IsEventSaved(createModel.Id);
-        if (IsEventInProcess(createModel.Id) || isSaved)
+        if (WhetherSaveEventBeingHandling(createModel.Id) || isSaved)
```
Почему Whether, а не Is...

DaGnYMilleR commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -104,10 +116,47 @@ public IActionResult GetEventsByName(string? eventName, [FromQuery(Name = "m")]
+        var photos = createModel.Photos;
+
+        if (!IsContentTypesSupported(photos))
+        {
```
зачем скобки?

DaGnYMilleR commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -104,10 +116,47 @@ public IActionResult GetEventsByName(string? eventName, [FromQuery(Name = "m")]
+
+        if (!IsPhotosSizeLessThanRecommended(photos))
+        {
+            var recommendedSizeInMb = rabbitMqConfig.RecommendedMessageSizeInBytes / 1024 / 1024;
```
зачем это считать при каждом вызове метода?

DaGnYMilleR commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -130,6 +179,35 @@ public async Task<IActionResult> DeleteEvent(Guid eventId)
+
+    private static bool IsContentTypesSupported(IFormFile[] files)
+    {
+        var supportedContentTypes = new [] { "image/bmp", "image/png", "image/jpeg" };
```
тоже стоит делать где-нить выше (хотяб в конструкторе, но лучше еще выше)

DaGnYMilleR commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -130,6 +179,35 @@ public async Task<IActionResult> DeleteEvent(Guid eventId)
+    private static bool IsContentTypesSupported(IFormFile[] files)
+    {
+        var supportedContentTypes = new [] { "image/bmp", "image/png", "image/jpeg" };
+        var contentTypes = files.Select(photo => photo.ContentType);
```
можно одним All обойтись бтв (но, кажется, что ненужно, так чиатемей)

DaGnYMilleR commented on Eventnet/Eventnet.Api/Controllers/EventController.cs:
```diff
@@ -130,6 +179,35 @@ public async Task<IActionResult> DeleteEvent(Guid eventId)
+    public IActionResult GenerateEventGuid()
+    {
+        var id = Guid.NewGuid();
+        return Ok(id);
```
Нет, это штука не должна возвращать id, он генериться именно при запросе создания. Тут должны быть проверки на то, может ли юзер создать евент (понятно, что пока этого нет, поэтому можно просто Ok() возвращать)

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/SaveServices/EventSaveService.cs:
```diff
@@ -0,0 +1,58 @@
+    public SaveEventResult GetSaveEventResult(Guid id)
+    {
+        if (!handler.TryGetValue(id, out var saveEventResult))
+            return new SaveEventResult(EventSaveStatus.NotSavedDueToUserError, "No such Guid");
```
зачем говорить, о том что у нас Guid, фронтендеры вообще не знают, что это такое, у них UUID вроде

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/SaveServices/EventSaveService.cs:
```diff
@@ -0,0 +1,58 @@
+
+public class EventSaveService : IEventSaveService
+{
+    private readonly Handler handler;
```
да почему Handler, уже какое ревью пишу об этом ептыть

DaGnYMilleR commented on Eventnet/Eventnet.Api/Services/SaveServices/RabbitMqConfig.cs:
```diff
@@ -0,0 +1,8 @@
+
+public class RabbitMqConfig
+{
+    public string Queue { get; set; }
```
![image](https://user-images.githubusercontent.com/37438377/163708404-682d5691-cbdd-47b5-aaac-dbf96e32604d.png)

gODeaLoAple replied:
Это же конфиг...

gODeaLoAple replied:
Хотя Suppress не помешал бы, да.

DaGnYMilleR commented on Eventnet/Domain/Domain.csproj:
```diff
@@ -0,0 +1,9 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
в Domain неймспейсы говна
![image](https://user-images.githubusercontent.com/37438377/163708431-fc1549b4-bcef-4857-abfc-fee4b1509e7b.png)

DaGnYMilleR commented on Eventnet/Eventnet.Infrastructure/RabbitMqMessageHandler.cs:
```diff
@@ -0,0 +1,52 @@
+        handler.Update(id, new SaveEventResult(status, errorMessage));
+    }
+
+    private List<Photo> GetPhotos(List<RabbitMqPhoto> rabbitMqPhotos) => 
```
![image](https://user-images.githubusercontent.com/37438377/163708491-bf719398-e68f-434e-82ee-3d29efade173.png)

IvanSibirtsev committed:
add #pragma warning disable for config classes

IvanSibirtsev committed:
rename Handler -> EventSaveHandler. Finally

IvanSibirtsev committed:
small fixes

IvanSibirtsev committed:
delete migrations

IvanSibirtsev committed:
Merge conflicts

IvanSibirtsev committed:
add AddCreateEvent migrations

IvanSibirtsev committed:
use CreateAuthorizedClient() in integration tests
